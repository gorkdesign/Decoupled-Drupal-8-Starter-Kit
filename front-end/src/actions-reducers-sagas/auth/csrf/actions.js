// ---------------
// Actions
// ---------------

export const INIT_CSRF_TOKEN = 'INIT_CSRF_TOKEN';
export const CSRF_TOKEN_SUCCESS = 'CSRF_TOKEN_SUCCESS';

export const SET_CSRF_TOKEN = 'SET_CSRF_TOKEN';

// ---------------
// Action Creators
// ---------------

//GETTERS
export const initCsrfToken = () => ({
  type: INIT_CSRF_TOKEN
});

export const csrfTokenSuccess = () => ({
  type: CSRF_TOKEN_SUCCESS
});

//SETTERS
export const setCsrfToken = (payload) => ({
  type: SET_CSRF_TOKEN,
  payload: payload
})
