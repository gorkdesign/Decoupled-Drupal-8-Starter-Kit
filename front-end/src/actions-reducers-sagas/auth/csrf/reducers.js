export const initialState = {
  csrfToken: null
};

export const reducer = (state = initialState, {type, payload}) => {
  switch (type) {

    case 'SET_CSRF_TOKEN':
      console.log("CSRF TOKEN SET");
      return {
        ...state,
        csrfToken: payload,
      };
    default:
      return state;
  }
};

export default reducer;
