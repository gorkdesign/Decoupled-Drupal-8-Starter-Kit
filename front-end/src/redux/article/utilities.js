export const generateImageUrlMid = (images) => {
  const newImages = images.map(image => (
    { url: image.entity.image.derivative.url, mid: image.mid }
  ));
  return newImages;
}

export const formatFetchArticlesResult = result => result.data.user.nodes.articles.map((node) => {
  const newNode = { ...node };
  newNode.images = generateImageUrlMid(node.images);
  return newNode;
});

export const removeArticleFromArticles = (articles, nid) => {
  const newArticles = articles.slice();
  const index = newArticles.findIndex(item => item.nid === nid);
  if (index === -1) { return false; }
  newArticles.splice(index, 1);
  return newArticles;
};

export const updateArticlesWithArticle = (articles, page) => {
  const newArticles = articles.slice();
  const index = newArticles.findIndex(item => item.nid === page.nid);
  if (index === -1) { return newArticles; }
  newArticles[index] = page;
  return newArticles;
};

export const getArticleFromNid = (articles, nid) => {
  const index = articles.findIndex(item => item.nid === nid);
  return articles[index];
};
