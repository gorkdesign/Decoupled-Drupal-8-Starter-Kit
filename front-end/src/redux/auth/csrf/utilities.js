import axios from 'axios';

const URL = process.env.REACT_APP_HOST_DOMAIN;

export const fetchCsrfToken = (url = URL) => axios.get(`${url}/session/token`)
  .then(response => (response.data))
  .catch(error => error);

export default {};
