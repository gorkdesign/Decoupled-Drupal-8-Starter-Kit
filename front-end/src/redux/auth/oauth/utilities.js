import Querystring from 'query-string';
import axios from 'axios';

const URL = process.env.REACT_APP_HOST_DOMAIN;
const POSTFIX = process.env.REACT_APP_XDEBUG_POSTFIX;

const CLIENT_INFO = {
  client_id: process.env.REACT_APP_CLIENT_ID,
  client_secret: process.env.REACT_APP_CLIENT_SECRET,
};

export const fetchToken = credentials => axios.post(`${URL}/oauth/token${POSTFIX}`, Querystring.stringify(credentials))
  .then((response) => {
    const { expires_in, access_token, refresh_token } = response.data; // eslint-disable-line
    return ({
      expiration: expires_in,
      accessToken: access_token,
      refreshToken: refresh_token,
      timestamp: new Date().getTime(),
    });
  }).catch((error) => {
    console.log(error);
    return error;
  });

export const generateCredentials = (type, username, password = '', refreshToken = '') => {
  let credentials = {
    ...CLIENT_INFO,
    grant_type: type,
    username,
  };

  if (type === 'password') {
    credentials = {
      ...credentials,
      password,
    };
  } else if (type === 'refresh_token') {
    credentials = {
      ...credentials,
      refresh_token: refreshToken,
    };
  }

  return credentials;
};

export const persistUsername = (username) => {
  sessionStorage.setItem('username', username);
};

export const persistCredentials = ({
  accessToken, expiration, refreshToken, timestamp,
}) => {
  sessionStorage.setItem('accessToken', accessToken);
  sessionStorage.setItem('expirationTime', expiration);
  localStorage.setItem('refreshToken', refreshToken);
  localStorage.setItem('lastRefreshedToken', timestamp);
};

export const getLocalCredentials = () => ({
  accessToken: sessionStorage.getItem('accessToken'),
  expireStamp: localStorage.getItem('lastRefreshedToken'),
  csrfToken: sessionStorage.getItem('csrfToken'),
  refreshToken: localStorage.getItem('refreshToken'),
});

export const unsetLocalStorageCredentials = () => {
  sessionStorage.removeItem('accessToken');
  sessionStorage.removeItem('username');
  sessionStorage.removeItem('csrfToken');
  sessionStorage.removeItem('expirationTime');
  localStorage.removeItem('refreshToken');
  localStorage.removeItem('lastRefreshedToken');
  localStorage.removeItem('state');
};

const convertDate = (unixTimestamp) => {
  // Create a new JavaScript Date object based on the timestamp
  // multiplied by 1000 so that the argument is in milliseconds, not seconds.
  const date = new Date(unixTimestamp);
  // Hours part from the timestamp
  const hours = date.getHours();
  // Minutes part from the timestamp
  const minutes = `0${date.getMinutes()}`;
  // Seconds part from the timestamp
  const seconds = `0${date.getSeconds()}`;

  // Will display time in 10:30:23 format
  const formattedTime = `${hours}:${minutes.substr(-2)}:${seconds.substr(-2)}`;
  return formattedTime;
};

export const isTokenValid = (accessToken, expiresStamp) => {
  const currentTime = new Date().getTime();
  const expireTime = parseInt(sessionStorage.getItem('expirationTime'), 10) * 1000;


  const currentTimeInt = parseInt(currentTime, 10);
  const expiresStampInt = parseInt(expiresStamp, 10) + expireTime;

  console.log(`CURRENT-TIME: ${convertDate(currentTimeInt)}`);
  console.log(`EXPIRES-TIME: ${convertDate(expiresStampInt)}`);

  if (accessToken && (expiresStampInt > currentTimeInt)) {
    return true;
  }
  return false;
};
