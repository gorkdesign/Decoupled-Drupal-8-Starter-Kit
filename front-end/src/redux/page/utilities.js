export const formatFetchPageResult = result => result.data.user.nodes.pages.map((node) => {
  const newNode = { ...node };
  newNode.images = newNode.images.map(image => (
    { url: image.entity.image.derivative.url, mid: image.mid }
  ));
  return newNode;
});

export const removePageFromPages = (pages, nid) => {
  const newPages = pages.slice();
  const index = newPages.findIndex(item => item.nid === nid);
  if (index === -1) { return false; }
  newPages.splice(index, 1);
  return newPages;
};

export const updatePagesWithPage = (pages, page) => {
  const newPages = pages.slice();
  const index = newPages.findIndex(item => item.nid === page.nid);
  if (index === -1) { return newPages; }
  newPages[index] = page;
  return newPages;
};

export const getPageFromNid = (pages, nid) => {
  const index = pages.findIndex(item => item.nid === nid);
  return pages[index];
};
