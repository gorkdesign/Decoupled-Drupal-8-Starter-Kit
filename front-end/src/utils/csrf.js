import axios from 'axios';

const URL = process.env.REACT_APP_HOST_DOMAIN;

function initializeCsrfToken(){
  axios.get(URL + '/session/token')
    .then(response => {
      sessionStorage.setItem('csrfToken', response.data);
    })
    .catch((error) => {
      console.log('error ' + error);
    });
};

const initalizeCsrfToken = () => {
  if(!sessionStorage.getItem('csrfToken') ){
    initializeCsrfToken();
  }
}

export default initalizeCsrfToken;