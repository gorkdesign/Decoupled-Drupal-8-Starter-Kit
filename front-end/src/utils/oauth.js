import axios from 'axios';
import Querystring from 'query-string';
import { getSessionStorage, clearSessionStorage } from './sessionStorage';

const POSTFIX = process.env.REACT_APP_XDEBUG_POSTFIX;
const URL = process.env.REACT_APP_HOST_DOMAIN;
const CLIENT_INFO = {
  client_id: process.env.REACT_APP_CLIENT_ID,
  client_secret: process.env.REACT_APP_CLIENT_SECRET
};

const isTokenValid = (accessToken, expiresIn) => {

  const currentTime = new Date().getTime();
  if(accessToken && (!expiresIn || expiresIn - currentTime > 1)){
    console.log('isTokenValid: (TRUE) TOKEN EXPIRES AT: ' + new Date(parseInt(expiresIn, 10)));
    return true;
  }else{
    console.log('isTokenValid: (FALSE) TOKEN EXPIRED');
    return false;
  }
}

const getRefreshToken = () => {
  const refreshToken = localStorage.getItem('refresh_token');
  if (refreshToken) {
    console.log('REFRESH TOKEN USED');
    return refreshToken;
  }else{
    return false;
  }
}

const getCredentials = (type, username, password = '', refreshToken = '') => {
  let credentials = {
    ...CLIENT_INFO,
    grant_type: type,
    username: username,
  }

  type === 'password' ? 
  credentials = {
    ...credentials,
    password: password
  }
  : 
  credentials = {
    ...credentials,
    refresh_token: refreshToken
  }

  return credentials;
}

const initializeOauthToken = (credentials, callback = (success, error) => {}) => {
  axios.post(URL + '/oauth/token' + POSTFIX, Querystring.stringify(credentials))
    .then(response => {
      const {expires_in, access_token, refresh_token} = response.data;
      const expiresIn = new Date().getTime() + expires_in * 1000;
      console.log('TOKEN EXPIRES AT: ' + new Date(parseInt(expiresIn, 10)));
      sessionStorage.setItem('access_token', access_token);
      sessionStorage.setItem('expires_in', expiresIn);
      localStorage.setItem('refresh_token', refresh_token);
      callback(true);
    })
    .catch((error) => {
      callback(false, error);
    });
};

export const getToken = () => {
  const { accessToken, expiresIn } = getSessionStorage();
  if(!accessToken || !expiresIn){
    return false;
  }

  if (isTokenValid(accessToken, expiresIn)) {
    return `Bearer ${accessToken}`;
  }
  
  return getRefreshToken();
}

export const authenticate = (username, password, callback=(success, error)=>{}) => {

  if(username){
    sessionStorage.setItem('username', username);
  }
  
  const { accessToken, expiresIn } = getSessionStorage();
  const refreshToken = !isTokenValid(accessToken, expiresIn) ? getRefreshToken() : undefined;


  if(accessToken && isTokenValid(accessToken, expiresIn)) {
    callback(true); 
    return;
  } 
  
  let credentials = (!accessToken) ? 
    getCredentials('password', username, password) : 
    getCredentials('refresh_token', username, password, refreshToken)

  initializeOauthToken(credentials, (success, error) => {
    if(!success){
      clearSessionStorage();
    }
    callback(success, error)
  });
}

export default authenticate;