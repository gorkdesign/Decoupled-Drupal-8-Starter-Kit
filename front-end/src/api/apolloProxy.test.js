import { ApolloLink, concat } from 'apollo-link';
import { HttpLink } from 'apollo-link-http';
import { ApolloClient } from 'apollo-client';
import { InMemoryCache, IntrospectionFragmentMatcher } from 'apollo-cache-inmemory';
import introspectionQueryResultData from 'api/fragmentTypes.json';

import {
  currentUserUidQuery,
  articlesByUserQuery,
  addArticleMutation,
  deleteArticleMutation,
  updateArticleMutation,
  getSignedUrls,
  addS3Files,
} from './apolloProxy';

import { fetchCsrfToken } from '../redux/auth/csrf/utilities';
import { fetchToken, generateCredentials } from '../redux/auth/oauth/utilities';

const POSTFIX = process.env.REACT_APP_XDEBUG_POSTFIX;
const URL = process.env.REACT_APP_HOST_DOMAIN ? process.env.REACT_APP_HOST_DOMAIN : '';

describe('ApolloProxy', async () => {
  const csrfToken = await fetchCsrfToken();
  const credentials = generateCredentials('password', 'admin', 'admin', '');
  const tokens = await fetchToken(credentials);

  const authMiddleware = new ApolloLink((operation, forward) => {
    const csrf = csrfToken;
    const token = tokens.accessToken;

    // add the access_token to the headers
    operation.setContext(() => ({
      headers: {
        authorization: `Bearer ${token}` || null,
        'X-CSRF-Token': csrf || null,
      },
    }));
    return forward(operation);
  });

  const link = new HttpLink({
    uri: URL.concat(`/graphql${POSTFIX}`),
  });

  const fragmentMatcher = new IntrospectionFragmentMatcher({ introspectionQueryResultData });

  const apolloClient = new ApolloClient({
    link: concat(authMiddleware, link),
    cache: new InMemoryCache({ fragmentMatcher }),
  });

  describe('Snapshots', () => {
    it('should render correctly', async () => {
      const userUidQuery = await apolloClient.query({ query: currentUserUidQuery });
      expect(userUidQuery).toMatchSnapshot();


      const articlesByUser = await apolloClient.query({ query: articlesByUserQuery });
      expect(articlesByUser).toMatchSnapshot();
    });
  });
});
