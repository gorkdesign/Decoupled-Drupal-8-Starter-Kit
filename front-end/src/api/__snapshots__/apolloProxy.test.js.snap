// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApolloProxy should render correctly 1`] = `
Object {
  "data": Object {
    "currentUserContext": Object {
      "__typename": "User",
      "uid": 1,
      "uuid": "4e9af657-689b-4c06-8721-e267914f2255",
      Symbol(id): "$ROOT_QUERY.currentUserContext",
    },
    Symbol(id): "ROOT_QUERY",
  },
  "loading": false,
  "networkStatus": 7,
  "stale": false,
}
`;

exports[`ApolloProxy should render correctly 2`] = `
Object {
  "data": Object {
    "user": Object {
      "__typename": "User",
      "nodes": Object {
        "__typename": "EntityQueryResult",
        "articles": Array [
          Object {
            "__typename": "NodeArticle",
            "author": Object {
              "__typename": "User",
              "name": "admin",
              Symbol(id): "$ROOT_QUERY.currentUserContext.reverseUidNode({\\"offset\\":0,\\"limit\\":1000}).entities.0.entityOwner",
            },
            "body": null,
            "images": Array [],
            "nid": 13,
            "title": "hello article update",
            "uuid": "79502776-61f8-4c48-b464-d94eebe0e01b",
            Symbol(id): "$ROOT_QUERY.currentUserContext.reverseUidNode({\\"offset\\":0,\\"limit\\":1000}).entities.0",
          },
          Object {
            "__typename": "NodeArticle",
            "author": Object {
              "__typename": "User",
              "name": "admin",
              Symbol(id): "$ROOT_QUERY.currentUserContext.reverseUidNode({\\"offset\\":0,\\"limit\\":1000}).entities.1.entityOwner",
            },
            "body": Object {
              "__typename": "FieldNodeBody",
              "value": "<p>asdfasdf</p>
",
              Symbol(id): "$ROOT_QUERY.currentUserContext.reverseUidNode({\\"offset\\":0,\\"limit\\":1000}).entities.1.body",
            },
            "images": Array [],
            "nid": 14,
            "title": "asdfads",
            "uuid": "be510cd5-645e-4f72-8baa-06cf89f53f84",
            Symbol(id): "$ROOT_QUERY.currentUserContext.reverseUidNode({\\"offset\\":0,\\"limit\\":1000}).entities.1",
          },
        ],
        Symbol(id): "$ROOT_QUERY.currentUserContext.reverseUidNode({\\"offset\\":0,\\"limit\\":1000})",
      },
      "uid": 1,
      Symbol(id): "$ROOT_QUERY.currentUserContext",
    },
    Symbol(id): "ROOT_QUERY",
  },
  "loading": false,
  "networkStatus": 7,
  "stale": false,
}
`;
